addons/mail/models/mail_followers.py

```diff
@@ -205,17 +205,27 @@ def _get_recipient_data(self, records, message_type, subtype_id, pids=None):
       JOIN sub_followers ON sub_followers.pid = partner.id
                         AND (sub_followers.internal IS NOT TRUE OR partner.partner_share IS NOT TRUE)
  LEFT JOIN LATERAL (
+        WITH RECURSIVE all_groups AS (
         SELECT users.id AS uid,
                users.share AS share,
                users.notification_type AS notification_type,
-               ARRAY_AGG(groups_rel.gid) FILTER (WHERE groups_rel.gid IS NOT NULL) AS groups
+               groups_rel.gid
           FROM res_users users
      LEFT JOIN res_groups_users_rel groups_rel ON groups_rel.uid = users.id
          WHERE users.partner_id = partner.id AND users.active
-      GROUP BY users.id,
-               users.share,
-               users.notification_type
-      ORDER BY users.share ASC NULLS FIRST, users.id ASC
+      UNION
+        SELECT ag.uid, ag.share, ag.notification_type, implied.hid
+        FROM all_groups ag
+        JOIN res_groups_implied_rel implied ON ag.gid = implied.gid)
+      SELECT uid,
+             share,
+             notification_type,
+          ARRAY_AGG(DISTINCT gid) AS groups
+      FROM all_groups
+      GROUP BY uid,
+               share,
+               notification_type
+      ORDER BY share ASC NULLS FIRST, uid ASC
```

Comment:
Mixing two different indentation conventions makes it kinda difficult to read.  I suggest to indent everything the SQL way (align on the right-hand side of keywords).
```suggestion
      UNION
        SELECT ag.uid, ag.share, ag.notification_type, implied.hid
          FROM all_groups ag
          JOIN res_groups_implied_rel implied ON ag.gid = implied.gid
        )
        SELECT uid,
               share,
               notification_type,
               ARRAY_AGG(DISTINCT gid) AS groups
          FROM all_groups
      GROUP BY uid,
               share,
               notification_type
      ORDER BY share ASC NULLS FIRST, uid ASC
```